### PR TITLE
Add batch inference

### LINKED ISSUE #25: batch inference
Hello,
I am trying to find a way to do batch inference with mmpose, but it seems that the project has not supported it yet. There is only a function called "**_inference_single_pose_model**" inside inference.py.

Is there any quick way I can adapt mmpose to batch inference? I think it's an essential feature because most of the time, we just work on bach inference. real-time infernece with a single input image is not possible with the current architectures, so single-image inference is kind of not useful.

Thank  you and I look forward to hearing from you soon.
Best.

### LINKED ISSUE #120: Multiple samples per GPU while testing
Hi everybody

Currently only one sample per GPU is possible while executing tests, which leads to very slow validation and testing for me. I assume to fix this the `forward_test` methods for the `TopDown` and `BottomUp` classes would need to be updated. Is this planned anytime soon or should I give it a try myself to contribute?

### LINKED ISSUE #197: Support Batch Inference
ref
```
https://github.com/open-mmlab/mmdetection/pull/3564
https://github.com/open-mmlab/mmdetection/pull/3686
https://github.com/open-mmlab/mmdetection/pull/3705
```
and
```
https://github.com/open-mmlab/mmaction2/pull/184/files
```

### PR DESCRIPTION
**Pull request by wusize**
Support batch inference for topdown methods.
fix #25 
fix #120 
fix #197

### CODE REVIEW

File: mmpose/models/detectors/top_down.py [hunk 1]
```diff
@@ -178,22 +178,22 @@ def forward_train(self, img, target, target_weight, img_metas, **kwargs):
 
     def forward_test(self, img, img_metas, return_heatmap=False, **kwargs):
         """Defines the computation performed at every call when testing."""
-        assert img.size(0) == 1
-        assert len(img_metas) == 1
-        img_metas = img_metas[0]
+        # assert img.size(0) == 1
```
Comment (jin-s13) [inline mmpose/models/detectors/top_down.py:181]: rm unnecessary comments.

File: mmpose/datasets/datasets/top_down/topdown_posetrack18_dataset.py [hunk 2]
```diff
@@ -191,6 +201,17 @@ def evaluate(self, outputs, res_folder, metric='mAP', **kwargs):
                 oks_nmsed_kpts[image_id].append(
                     [img_kpts[_keep] for _keep in keep])
 
+            if self.use_nms:
```
Comment (jin-s13) [inline mmpose/datasets/datasets/top_down/topdown_posetrack18_dataset.py:204]: Seems not related to this PR?

File: mmpose/datasets/datasets/top_down/topdown_mpii_trb_dataset.py [hunk 3]
```diff
@@ -230,3 +236,11 @@ def _report_metric(self, res_file):
         info_str.append(('Contour_acc', contour.item()))
         info_str.append(('PCKh', mean.item()))
         return info_str
+
+    def _drop_repeated(self, kpts):
+        num = len(kpts)
+        for i in range(num - 1, 0, -1):
```
Comment (jin-s13) [inline mmpose/datasets/datasets/top_down/topdown_mpii_trb_dataset.py:242]: This func. seems to be erroneous

File: mmpose/datasets/datasets/top_down/topdown_coco_dataset.py [hunk 4]
```diff
@@ -284,8 +291,10 @@ def _load_coco_person_detection_results(self):
                 'bbox_score': score,
                 'dataset': self.dataset_name,
                 'joints_3d': joints_3d,
-                'joints_3d_visible': joints_3d_visible
+                'joints_3d_visible': joints_3d_visible,
+                'bbox_id': num_boxes
             })
+            num_boxes = num_boxes + 1
```
Comment (jin-s13) [inline mmpose/datasets/datasets/top_down/topdown_coco_dataset.py:297]: `num_boxes` or `num`?

**Comment by jin-s13** [comment]
unittest fails

File: tests/test_datasets/test_top_down_dataset.py [hunk 5]
```diff
@@ -28,19 +28,21 @@ def load_json_to_output(json_name, prefix=''):
 
 def convert_db_to_output(db):
     outputs = []
-
+    bbox_id = 0
     for item in db:
         keypoints = item['joints_3d'].reshape((1, -1, 3))
         center = item['center']
         scale = item['scale']
+        bbox_ids = [bbox_id]
```
Comment (jin-s13) [inline tests/test_datasets/test_top_down_dataset.py:36]: Do we really need list?

File: mmpose/datasets/datasets/top_down/topdown_posetrack18_dataset.py (hunk 2 above)
Comment (wusize) [reply mmpose/datasets/datasets/top_down/topdown_posetrack18_dataset.py:204]: Em, this is not related to batch inference. Should I open another proposal for this? Or maybe just merge it.

File: tests/test_datasets/test_top_down_dataset.py (hunk 5 above)
Comment (wusize) [reply tests/test_datasets/test_top_down_dataset.py:36]: Yes, it must be a list and the length of the list would equal to the batch size.

File: mmpose/datasets/datasets/top_down/topdown_posetrack18_dataset.py (hunk 2 above)
Comment (jin-s13) [reply mmpose/datasets/datasets/top_down/topdown_posetrack18_dataset.py:204]: Just leave it here for now.

File: mmpose/datasets/datasets/top_down/topdown_coco_dataset.py [hunk 6]
```diff
@@ -359,14 +371,15 @@ def evaluate(self, outputs, res_folder, metric='mAP', **kwargs):
...
-            nms = soft_oks_nms if self.soft_nms else oks_nms
-            keep = nms(list(img_kpts), oks_thr, sigmas=self.sigmas)
-
-            if len(keep) == 0:
-                oks_nmsed_kpts.append(img_kpts)
+            if self.use_nms:
+                nms = soft_oks_nms if self.soft_nms else oks_nms
+                keep = nms(list(img_kpts), oks_thr, sigmas=self.sigmas)
+                if len(keep) == 0:
+                    oks_nmsed_kpts.append(img_kpts)
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_coco_dataset.py:378]: If img_kpts is non empty, then the result of nms could still be empty, right or wrong?

File: mmpose/datasets/datasets/top_down/topdown_coco_dataset.py [hunk 7]
```diff
@@ -359,14 +371,15 @@ def evaluate(self, outputs, res_folder, metric='mAP', **kwargs):
                 # rescoring
                 n_p['score'] = kpt_score * box_score
 
-            nms = soft_oks_nms if self.soft_nms else oks_nms
-            keep = nms(list(img_kpts), oks_thr, sigmas=self.sigmas)
-
-            if len(keep) == 0:
-                oks_nmsed_kpts.append(img_kpts)
+            if self.use_nms:
+                nms = soft_oks_nms if self.soft_nms else oks_nms
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_coco_dataset.py:375]: I'm thinking, whether we could merge `self.soft_nms` and `self.use_nms` to something like `self.nms_method=none|oks|soft`

File: mmpose/datasets/datasets/top_down/topdown_coco_dataset.py [hunk 8]
```diff
@@ -435,3 +448,13 @@ def _do_python_keypoint_eval(self, res_file):
         info_str = list(zip(stats_names, coco_eval.stats))
 
         return info_str
+
+    def _drop_repeated(self, kpts, key='bbox_id'):
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_coco_dataset.py:452]: docstring for this function

File: mmpose/datasets/datasets/top_down/topdown_coco_dataset.py [hunk 9]
```diff
@@ -435,3 +448,13 @@ def _do_python_keypoint_eval(self, res_file):
...
+        for img_id, bboxes in kpts.items():
+            num = len(bboxes)
+            kpts[img_id] = sorted(kpts[img_id], key=lambda x: x[key])
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_coco_dataset.py:455]: Is the order of kpts essential for this function?

File: mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py [hunk 10]
```diff
@@ -45,9 +45,15 @@ def __init__(self,
...
+                 test_mode=False,
+                 use_nms=True):
         super(TopDownCocoDataset, self).__init__(
-            ann_file, img_prefix, data_cfg, pipeline, test_mode=test_mode)
+            ann_file,
+            img_prefix,
+            data_cfg,
+            pipeline,
+            test_mode=test_mode,
+            use_nms=use_nms)
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py:56]: Put use_nms under  dataset construction feels strange. How about making it an argument for evaluation?

File: mmpose/datasets/datasets/top_down/topdown_mpii_dataset.py [hunk 11]
```diff
@@ -147,12 +152,12 @@ def evaluate(self, outputs, res_folder, metric='PCKh', **kwargs):
...
 
-                * preds(np.ndarray[1,K,3]): The first two dimensions are
+                * preds(np.ndarray[N,K,3]): The first two dimensions are
                   coordinates, score is the third dimension of the array.
-                * boxes(np.ndarray[1,6]): [center[0], center[1], scale[0]
+                * boxes(np.ndarray[N,6]): [center[0], center[1], scale[0]
                   , scale[1],area, score]
-                * image_path(list[str]): For example, ['0', '0',
-                  '0', '0', '0', '1', '1', '6', '3', '.', 'j', 'p', 'g']
+                * image_path(list[str]): For example, ['data/coco/val2017
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_mpii_dataset.py:159]: `image_path` -> `image_paths`?

File: mmpose/datasets/datasets/top_down/topdown_coco_dataset.py (hunk 9 above)
Comment (innerlee) [reply mmpose/datasets/datasets/top_down/topdown_coco_dataset.py:455]: If so, then the function name could be `_sort_and_unique_bboxes`

If not, you can use
```python
kpts[img_id] = list({bbox[key]: bbox for bbox in bboxes}.values())
```

File: mmpose/datasets/datasets/top_down/topdown_posetrack18_dataset.py (hunk 2 above)
Comment (innerlee) [reply mmpose/datasets/datasets/top_down/topdown_posetrack18_dataset.py:204]: Split independent changes to separate prs next time

File: tests/test_datasets/test_top_down_dataset.py [hunk 12]
```diff
@@ -26,21 +26,28 @@ def load_json_to_output(json_name, prefix=''):
...
+    len_db = len(db)
+    for i in range(0, len_db, batch_size):
+        keypoints = np.stack([
+            db[j]['joints_3d'].reshape((-1, 3))
+            for j in range(i, min(i + batch_size, len_db))
+        ])
+        img_path = [
+            db[j]['image_file'] for j in range(i, min(i + batch_size, len_db))
+        ]
+        bbox_ids = [j for j in range(i, min(i + batch_size, len_db))]
```
Comment (innerlee) [inline tests/test_datasets/test_top_down_dataset.py:40]:
```python
list(range(i, min(i + batch_size, len_db)))
```

**Comment by innerlee** [comment]
Overall, the change is huge for a small functionality. 🤯

File: mmpose/datasets/datasets/top_down/topdown_coco_dataset.py (hunk 6 above)
Comment (wusize) [reply mmpose/datasets/datasets/top_down/topdown_coco_dataset.py:378]: The result of nms won't be empty as long as img_kpts has at least one element in it.  But this operation has been there in previous versions of mmpose, so I just kept it. 
May I remove these lines?

File: mmpose/datasets/datasets/top_down/topdown_coco_dataset.py (hunk 7 above)
Comment (wusize) [reply mmpose/datasets/datasets/top_down/topdown_coco_dataset.py:375]: Yes, we can. And this would require deprecation warnings if we were not to change the config files of the models.

File: mmpose/datasets/datasets/top_down/topdown_coco_dataset.py (hunk 8 above)
Comment (wusize) [reply mmpose/datasets/datasets/top_down/topdown_coco_dataset.py:452]: OK

File: mmpose/datasets/datasets/top_down/topdown_mpii_dataset.py (hunk 11 above)
Comment (wusize) [reply mmpose/datasets/datasets/top_down/topdown_mpii_dataset.py:159]: Ok.

File: mmpose/datasets/datasets/top_down/topdown_coco_dataset.py (hunk 9 above)
Comment (wusize) [reply mmpose/datasets/datasets/top_down/topdown_coco_dataset.py:455]: OK. I should rename the function.

File: mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py (hunk 10 above)
Comment (wusize) [reply mmpose/datasets/datasets/top_down/topdown_crowdpose_dataset.py:56]: I'm thinking about putting 'use_nms' in **data_cfg** in config files of models, given that args like 'soft_nms' , 'nms_thr', 'oks_thr', 'vis_thr' and 'bbox_thr' are all specified in **data_cfg**. And since different evaluation methods are often unique to different datasets we used, it someway makes sense to put these args under dataset construction.

File: mmpose/datasets/datasets/top_down/topdown_posetrack18_dataset.py (hunk 2 above)
Comment (wusize) [reply mmpose/datasets/datasets/top_down/topdown_posetrack18_dataset.py:204]: Em, I'll open another  pr to add **use_nms**.

**Comment by akhalili** [comment]
When do you guys target to add this feature to master?

Thanks

**Comment by innerlee** [comment]
This will be refactored, so that users do not need to change configs

**Comment by innerlee** [comment]
kindly ping @wusize
